flows/engine/session.go

```diff
@@ -51,117 +60,302 @@ func (s *session) GetRun(uuid flows.RunUUID) (flows.FlowRun, error) {
 	return nil, fmt.Errorf("unable to find run with UUID: %s", uuid)
 }
 
-func (s *session) ActiveRun() flows.FlowRun {
-	for _, run := range s.runs {
-		// We are complete, therefore can't be active
-		if run.IsComplete() {
-			continue
-		}
+func (s *session) addRun(run flows.FlowRun) {
+	s.runs = append(s.runs, run)
+	s.runsByUUID[run.UUID()] = run
+}
 
-		// We have a child, and it isn't complete, we can't be active
-		if run.Child() != nil && run.Child().Status() == flows.StatusActive {
-			continue
-		}
+func (s *session) Status() flows.SessionStatus { return s.status }
+func (s *session) Wait() flows.Wait            { return s.wait }
 
-		return run
+// looks through this session's run for the one that is waiting
+func (s *session) waitingRun() flows.FlowRun {
+	for _, run := range s.runs {
+		if run.Status() == flows.RunStatusWaiting {
+			return run
+		}
 	}
 	return nil
 }
 
-func (s *session) addRun(run flows.FlowRun) {
-	s.runs = append(s.runs, run)
-	s.runsByUUID[run.UUID()] = run
-}
-
 func (s *session) LogEvent(step flows.Step, action flows.Action, event flows.Event) {
 	s.log = append(s.log, NewLogEntry(step, action, event))
 }
 func (s *session) Log() []flows.LogEntry { return s.log }
 func (s *session) ClearLog()             { s.log = nil }
 
-// StartFlow starts the flow for the passed in contact, returning the created FlowRun
-func (s *session) StartFlow(flowUUID flows.FlowUUID, parent flows.FlowRun, callerEvents []flows.Event) error {
-	flow, err := s.assets.GetFlow(flowUUID)
+//------------------------------------------------------------------------------------------
+// Flow execution
+//------------------------------------------------------------------------------------------
+
+// StartFlow beings execution of the given flow in this session
+func (s *session) StartFlow(flowUUID flows.FlowUUID, callerEvents []flows.Event) error {
+	flow, _ := s.Assets().GetFlow(flowUUID)
```

Comment:
No will fix. I had played with starting the main flow with a trigger_flow event, and in that case I think this sort of check would be done at the request (assets>session>events) validation stage... but starting the main flow with an event is tricky because we don't currently support events without runs - i.e. there's no step to attach them to.